grpc/read_request.cpp

```diff
@@ -0,0 +1,151 @@
+#include "read_request.hpp"
+
+
+namespace grpc_dnet {
+
+read_request_manager::read_request_manager(service_provider_info *provider_in, THandler handler_in)
+	: request_manager_base(provider_in)
+	, rpc_call_io_(&ctx_)
+	, response_json_offset_(0)
+	, response_data_offset_(0)
+	, handler_(handler_in)
+{
+	on_complete_op_ = on_complete_op::HANDLE_REQUEST;
+	provider_->service->RequestRead(&ctx_, &rpc_request_, &rpc_call_io_, provider_->cq, provider_->cq, this);
+}
+
+
+void read_request_manager::send_response(const std::shared_ptr<TResponse> &response) {
+	response_ = response;
+	on_complete_op_ = send_next();
+}
+
+
+void read_request_manager::process_completion(bool) {
+	switch (on_complete_op_) {
+	case on_complete_op::HANDLE_REQUEST:
+		new read_request_manager(provider_, handler_);
+		handle_request();
+		break;
+
+	case on_complete_op::HANDLE_SEND_PARTIAL_COMPLETE:
+		on_complete_op_ = send_next();
+		break;
+
+	case on_complete_op::HANDLE_SEND_ALL_COMPLETE:
+		send_finish_marker();
+		on_complete_op_ = on_complete_op::DELETE;
+		break;
+
+	case on_complete_op::DELETE:
+		delete this;
+		break;
+	}
+}
+
+
+void read_request_manager::handle_request() {
+	auto request = std::make_shared<TRequest>();
```

Comment:
`shared_ptr` is temporary solution.Then whole code became compilable, we change it to `unique_ptr` and polish data owning scheme. Now `unique_ptr` can cause problems because `std::function` doesn't work properly with movable-only objects (at least in C++11). `folly::Function` works ok, but we don't use folly in this project. I think firstly we must get working code, and only when polish. I agree that finally here must be `unique_ptr`.

Reply:
```
#include <iostream>
#include <memory>
#include <functional>

using namespace std;
using THandler = function<void (unique_ptr<int>&&)>;

int main() {
    auto a = make_unique<int>(1);
    
    THandler f = [](unique_ptr<int>&& x) {
        cout << *x << endl;
    };
    
    f(std::move(a));
}
```